mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,964 @@
+    Operation *srcOp = loadAndStoreOps[i];
+    MemRefAccess srcAccess(srcOp);
+    for (unsigned j = i + 1; j < numOps; ++j) {
+      Operation *dstOp = loadAndStoreOps[j];
+      MemRefAccess dstAccess(dstOp);
+      if (srcOp == dstOp)
+        continue;
+
+      // Criteria 2
+      if (srcAccess.memref == dstAccess.memref) {
+        // For two memref accesses Ax+B, matrix A of both accesses should be 
+        // equal. Only B should differ in last index.
+        if (this->loopAccessMatrices[srcOp] != this->loopAccessMatrices[dstOp])
+          continue;
+
+        SmallVector<int64_t, 4> srcOpCV = this->constVector[srcOp];
+        SmallVector<int64_t, 4> destOpCV = this->constVector[dstOp];
+        bool onlyLastIndexVaries = true;
+
+        for (unsigned i = 0; i < srcOpCV.size() - 1; i++) {
+          if ((srcOpCV[i] != destOpCV[i])) {
+            onlyLastIndexVaries = false;
+            break;
+          }
+        }
+
+        if (!onlyLastIndexVaries)
+          continue;
+
+        // Difference in values at last index should be less than
+        // cache_line_size/elementSize for a useful locality.
+        unsigned elementSize = this->elementSize[srcOp];
+        if (!(abs(srcOpCV[srcOpCV.size() - 1] -
+                  destOpCV[destOpCV.size() - 1]) <=
+              cache_line_size / elementSize))
+          continue;
+
+        // Insert `dstOp` into the group of `srcOp`.
+        insertIntoReferenceGroup(referenceGroups, groupId, srcOp, dstOp);
+
+      } else {
+        // Criteria 1
+        for (unsigned depth = 1; depth <= maxDepth + 1; depth++) {
+          FlatAffineConstraints dependenceConstraints;
+          SmallVector<DependenceComponent, 2> depComps;
+          DependenceResult result = checkMemrefAccessDependence(
+              srcAccess, dstAccess, depth, &dependenceConstraints, &depComps);
+
+          if (!hasDependence(result))
+            continue;
+          // Criteria 1.1
+          if (depth == maxDepth + 1) {
+            // Loop-independent dependence. Put both instructions in same
+            // reference group.
+            insertIntoReferenceGroup(referenceGroups, groupId, srcOp, dstOp);
+          } else {
+            // Criteria 1.2
+            // Search for dependence at depths other than innermostIndex.
+            // All entries should be zero.
+            bool hasDependency = false;
+            for (unsigned i = 0; i <= maxDepth; i++) {
+              if ((depComps[i].lb.getValue() != 0) &&
+                  (i != innermostIndex)) {
+                hasDependency = true;
+                break;
+              }
+            }
+            if (hasDependency)
+              continue;
+            else if (abs(depComps[innermostIndex]
+                             .lb.getValue()) <= 2)
+              insertIntoReferenceGroup(referenceGroups, groupId, srcOp, dstOp);
+          }
+        }
+      }
+    }
+  }
+  this->referenceGroups = referenceGroups;
```

Comment:
You shouldn't really be storing state this way. Can you make the function take an output parameter?

Reply:
Removed this one and similar others. No need to use output parameter, since this is a member function, it can access protected members directly.

Reply:
Update: Removed use of all state variables.